src/Compilers/CSharp/Portable/Binder/SwitchExpressionBinder.cs

```diff
@@ -0,0 +1,148 @@
+            {
+                diagnostics.Add(ErrorCode.ERR_SwitchExpressionNoBestType, SwitchExpressionSyntax.Location);
+                commonType = CreateErrorType();
+            }
+
+            return commonType;
+        }
+
+        /// <summary>
+        /// Rewrite the expressions in the switch expression cases to add a conversion to the result (common) type.
+        /// </summary>
+        private ImmutableArray<BoundSwitchExpressionCase> AddConversionsToCases(ImmutableArray<BoundSwitchExpressionCase> switchCases, TypeSymbol resultType, DiagnosticBag diagnostics)
+        {
+            var builder = ArrayBuilder<BoundSwitchExpressionCase>.GetInstance();
+            foreach (var oldCase in switchCases)
+            {
+                var oldValue = oldCase.Value;
+                var newValue = GenerateConversionForAssignment(resultType, oldValue, diagnostics);
+                var newCase = (oldValue == newValue) ? oldCase :
+                    new BoundSwitchExpressionCase(oldCase.Syntax, oldCase.Locals, oldCase.Pattern, oldCase.Guard, newValue, oldCase.HasErrors);
+                builder.Add(newCase);
+            }
+
+            return builder.ToImmutableAndFree();
+        }
+
+        private ImmutableArray<BoundSwitchExpressionCase> BindSwitchExpressionCases(SwitchExpressionSyntax node, Binder originalBinder, DiagnosticBag diagnostics)
+        {
+            bool hasErrors = SwitchGoverningExpression.HasErrors;
+            var builder = ArrayBuilder<BoundSwitchExpressionCase>.GetInstance();
+            foreach (var arm in node.Arms)
+            {
+                var armBinder = originalBinder.GetBinder(arm);
+                var boundArm = armBinder.BindSwitchExpressionArm(arm, diagnostics);
+                builder.Add(boundArm);
+            }
+
+            return builder.ToImmutableAndFree();
+        }
+
+        internal BoundExpression SwitchGoverningExpression
+        {
+            get
+            {
+                if (_switchGoverningExpression == null)
+                {
+                    EnsureSwitchGoverningExpressionAndDiagnosticsBound();
+                }
+
+                Debug.Assert(_switchGoverningExpression != null);
+                return _switchGoverningExpression;
+            }
+        }
+
+        internal TypeSymbol SwitchGoverningType => SwitchGoverningExpression.Type;
+
+        protected DiagnosticBag SwitchGoverningDiagnostics
+        {
+            get
+            {
+                EnsureSwitchGoverningExpressionAndDiagnosticsBound();
+                return _switchGoverningDiagnostics;
+            }
+        }
+
+        private void EnsureSwitchGoverningExpressionAndDiagnosticsBound()
+        {
+            var switchGoverningDiagnostics = new DiagnosticBag();
+            var boundSwitchGoverningExpression = BindSwitchGoverningExpression(switchGoverningDiagnostics);
+            _switchGoverningDiagnostics = switchGoverningDiagnostics;
```

Comment:
There doesn't need to be. Both values (old and new) will be semantically identical. We only need to ensure we have one of the computed diagnostic bags stored when we make _switchGoverningExpression non-null.

---
In reply to: [166481474](https://github.com/dotnet/roslyn/pull/23357#discussion_r166481474) [](ancestors = 166481474)

Reply:
After looking at this more, because we never check to see if the value is non-null before binding, why have a field at all? It looks like this recomputes the same information on every call to SwitchGoverningDiagnostics.

Reply:
We do check in the caller `SwitchGoverningExpression`, but not in the other caller. I'll move the check here. Now fixed in https://github.com/dotnet/roslyn/pull/23921/commits/6aafbf182d784bac51b870ae9e09c3d277e4c4ae of https://github.com/dotnet/roslyn/pull/23921.

Reply:
It looks like there still be a race condition. Here's my thinking:

1. Thread (1) accesses SwitchGoverningExpression.
2. Thread (1) proceeds through EnsureSwitchGoverningExpressionAndDiagnosticBound in completes binding.
3. The assignment to `_switchGoverningDiagnostics` and the cmpxchg are re-ordered such that the cmpxchg happens first. As far as I can see, this is legal according to the memory model.
4. The cmpxchg completes.
5. Thread (1) is suspended.
6. Thread (2) comes in and accesses `SwitchGoverningDiagnostics`, which skips over `EnsureSwitchGoverningExpressionAndDiagnosticsBound` and returns `null`.

Reply:
As I understand it, all writes preceding a compare-exchange are required to be flushed beforehand. This is a much stronger memory model that provided by Java.

If you believe that is wrong, can you please provide a reference?

Reply:
I re-read the .NET 2.0 memory model spec. I think you're right -- the barrier seems to publish to all reads and is a full barrier. I think this should be OK, then.